Implement formdata event

CYBAI opened the pull request:
---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #22654 
- [x] There are tests for these changes

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/22660)
<!-- Reviewable:end -->

CYBAI committed:
Implement formdata event

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/formdataevent.rs, components/script/dom/htmlformelement.rs, components/script/dom/mod.rs, components/script/dom/webidls/FormDataEvent.webidl, components/script/dom/formdata.rs
 * @kichjang: components/script/dom/formdataevent.rs, components/script/dom/htmlformelement.rs, components/script/dom/mod.rs, components/script/dom/webidls/FormDataEvent.webidl, components/script/dom/formdata.rs

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/web-platform-tests/wpt/pull/14768.

CYBAI commented on tests/wpt/web-platform-tests/xhr/formdata.htm:
```diff
@@ -56,8 +56,8 @@
-    assert_equals(formData.get('h1'), 'vh1');
-    assert_equals(formData.get('n1'), 'v1');
+    assert_equals(formDataInEvent.get('h1'), 'vh1');
+    assert_equals(formDataInEvent.get('n1'), 'v1');
```
I cannot pass the original test case locally and then I found it might be more reasonable to test with `formDataInEvent` here? because it's assigned with the appended `e.formData` which should have `h1`.

jdm replied:
This change should not be necessary if my earlier comment is addressed.

jdm commented on components/script/dom/htmlformelement.rs:
```diff
@@ -332,30 +343,36 @@ impl HTMLFormElement {
+            None => return,
+        };
+
+        // TODO: Step 10. If form cannot navigate, then return.
```
Could you implement this check? The algorithm at https://html.spec.whatwg.org/multipage/links.html#cannot-navigate is straightforward, and there are three uses of it in the form submissions algorithm.

CYBAI replied:
Sure! Let me try to implement it!

CYBAI replied:
Implemented in 038c83b

jdm commented on components/script/dom/formdataevent.rs:
```diff
@@ -0,0 +1,97 @@
+            Some(ref form_data) => form_data.clone(),
+            None => {
+                return Err(Error::Type(
+                    "required member formdata is undefined".to_string(),
```
Let's use `formData` in this string.

jdm commented on components/script/dom/formdataevent.rs:
```diff
@@ -0,0 +1,97 @@
+        }
+    }
+
+    pub fn new_uninitialized(global: &GlobalScope, form_data: &FormData) -> DomRoot<FormDataEvent> {
```
No need for this function; there are no callers so we can inline it in `new` instead.

jdm commented on components/script/dom/formdata.rs:
```diff
@@ -62,7 +62,15 @@ impl FormData {
         form: Option<&HTMLFormElement>,
     ) -> Fallible<DomRoot<FormData>> {
         // TODO: Construct form data set for form if it is supplied
```
We can remove this TODO. Also let's add a link to https://xhr.spec.whatwg.org/#dom-formdata.

CYBAI replied:
Thanks for checking this one! I was confused if I should remove it or not!

jdm commented on components/script/dom/webidls/FormDataEvent.webidl:
```diff
@@ -0,0 +1,14 @@
+};
+
+dictionary FormDataEventInit : EventInit {
+  /*required*/ FormData formData;
```
What happens if we uncomment this? If there's an error, do we have an issue filed for it, and can we add a comment?

CYBAI replied:
It will fail with this compilation error. I didn't find a corresponding issue. I'll try to read the webidl codegen to see what the root cause is later.

```rust
error[E0599]: no function or associated item named `empty` found for type `dom::bindings::codegen::Bindings::FormDataEventBinding::FormDataEventInit` in the current scope
   --> /Users/cybai/codespace/mozilla/servo/target/debug/build/script-59d4418593395066/out/Bindings/FormDataEventBinding.rs:857:95
    |
259 | pub struct FormDataEventInit {
    | ---------------------------- function or associated item `empty` not found for this
...
857 |             crate::dom::bindings::codegen::Bindings::FormDataEventBinding::FormDataEventInit::empty()
    |             ----------------------------------------------------------------------------------^^^^^
    |             |
    |             function or associated item not found in `dom::bindings::codegen::Bindings::FormDataEventBinding::FormDataEventInit`
```

jdm replied:
Ok, it is clear to me why this happens now. Filed https://github.com/servo/servo/issues/22670 about it.

CYBAI replied:
Thanks a lot! After reading the issue, I can understand why it failed to compile with `required` :)

jdm reviewed (changes requested):
Great work! These changes were easy to read and understand.

CYBAI commented on components/script/dom/htmlformelement.rs:
```diff
@@ -704,8 +728,28 @@ impl HTMLFormElement {
+        let window = window_from_node(self);
+
+        // Step 6
+        let form_data = FormData::new(Some(ret.clone()), &window.global());
```
I can udnerstand why the testcase failed now but I'm still trying to figure out how to avoid the `clone` here 🤔 Would it be a right way to pass the first argument of `FormData::new` with `Option<&Vec<FormDatum>>`?

jdm replied:
I expect you want to pass `ret` itself, then add an API to FormData to extract it again later and return that value.

CYBAI replied:
Ah! Yeah! I was thinking from constructor perspective 😂 That must be a good solution to extract it again! Thanks!

CYBAI replied:
Just found there's an existing API to extract the datums.

https://github.com/servo/servo/blob/192fbafd11126744a299558d14c4b2c63e3cd2ca/components/script/dom/formdata.rs#L197-L204

One more question for this; not sure why there're two `impl FormData` but maybe it's better to combine them? 🤔

jdm replied:
Combining them is fine.

servo-wpt-sync commented:
No upstreamable changes; closed existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#14768.

CYBAI commented:
Just updated some easy fix and remove the `update test case` commit! Will keep working on addressing other comments!

CYBAI committed:
Implement cannot navigate

CYBAI commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 038c83b97f76111c68506af121e9b2a245575808 with merge d10915ab4641906ccf33a5650313d298fc7b3e43...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11418)

jdm commented on components/script/dom/element.rs:
```diff
@@ -3230,6 +3230,19 @@ impl Element {
+        // Step 1.
+        !document.is_fully_active() || (
+            // Step 2.
+            self.upcast::<Node>().type_id() != NodeTypeId::Element(ElementTypeId::HTMLElement(
```
`!self.is::<HTMLAnchorElement>()` is easier to read :)

CYBAI replied:
Yeah! This is better!

jdm commented on components/script/dom/element.rs:
```diff
@@ -3230,6 +3230,19 @@ impl Element {
+            // Step 2.
+            self.upcast::<Node>().type_id() != NodeTypeId::Element(ElementTypeId::HTMLElement(
+                HTMLElementTypeId::HTMLAnchorElement,
+            )) && !self.upcast::<Element>().is_connected()
```
This is a method on Element, so we should be able to call `self.is_connected()`

CYBAI replied:
Ah, I forgot to update it! Thanks for catching this 😂

jdm reviewed (changes requested):
Looking good!

CYBAI commented on tests/wpt/metadata/html/semantics/forms/form-submission-0/constructing-form-data-set.html.ini:
```diff
@@ -10,13 +10,3 @@
 
   [Entries added to "formData" IDL attribute should be submitted.]
```
This test case still gets failure.

https://github.com/servo/servo/blob/4ac5b8a3aea5df3bf374dcfdb22c34f5ff0bafe8/tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/constructing-form-data-set.html#L98-L109

I tried to add logs to check if the event is dispatched and it's triggered correctly.
So I'm wondering if the `search` is not updated correctly after submitting the form.

CYBAI replied:
Ok. I've found what exact the wrong output is, will keep investigating the root cause.

The expected search is `n1=v1&h1=vh1` but we got `h1=vh1&n1=v1`.

CYBAI replied:
I've observed one issue here but the test case still got failure.

About the observed issue is, we're using `BTreeMap` for `FormData` so when we do iteration for it, the data will be sorted automatically. I'm wondering maybe we need to change to `IndexMap` instead?

https://github.com/servo/servo/blob/4d94692b043479ef420c51eea2b3fc9e7e709232/components/script/dom/formdata.rs#L28

After observed above issue, I tried to use `IndexMap` but I found the test case still got failure. Then I tried to make a break point in this line

https://github.com/servo/servo/blob/4d94692b043479ef420c51eea2b3fc9e7e709232/components/script/dom/htmlformelement.rs#L471

```
(lldb) p *load_data.url.__0.ptr.pointer
(alloc::sync::ArcInner<url::Url>) $1 = {
  strong = {
    v = (value = 1)
  }
  weak = {
    v = (value = 1)
  }
  data = {
    serialization = {
      vec = {
        buf = {
          ptr = (pointer = "http://web-platform.test:8000/common/blank.html?n1=v1&h1=vh1", _marker = core::marker::PhantomData<unsigned char> @ 0x00007fbc259b51f0)
          cap = 94
        }
        len = 60
      }
    }
    scheme_end = 4
    username_end = 7
    host_start = 7
    host_end = 24
    host = {}
    port = {}
    path_start = 29
    query_start = {}
    fragment_start = {}
  }
}
```

I found the url was shown correctly with `?n1=v1&h1=vh1`.

But, when I added a log `console.log(iframe.contentWindow.location.href);` in the `iframe.onload` event listener. It shows the `href` is `about:blank` so the `location.search` would be empty string which will not be equal to `n1=v1&h1=vh1`.

I'm wondering if the onload event is triggered after the form submitted 🤔

jdm replied:
Ok, I did some investigation around iframe load events (which caused me to read through https://github.com/whatwg/html/issues/490 and remember a bunch of awful details). As far as I can tell, the reason we observe the about:blank behaviour in this test is that there is a load event queued for the iframe immediately when it is added to the document (for the default about:blank browsing context). I created a test to verify this theory:
```html
<script src="/resources/testharness.js"></script>
<script src="/resources/testharnessreport.js"></script>
<iframe name="i" id="i"></iframe>
<form id="f" action="/common/blank.html" target="i"></form>
<script>
  async_test(function(t) {
    var i = document.getElementById('i');
    i.onload = t.step_func(function() {
      if (i.contentWindow.location.href == "about:blank") { return; } 
      assert_equals(i.contentWindow.location.pathname, "/common/blank.html"); 
      t.done();
    });
    document.getElementById('f').submit();
  });
</script>
```
This passes in Servo for me today. The spec says that the iframe load event should be queued when it's added to the document, and that form submission causes a task to be queued (and that task initiates a navigation for the iframe's browsing context), so it seems perfectly valid for the about:blank's load event to fire during that time before the iframe has officially been navigated away from about:blank.

THEREFORE

I think the right solution here is to add a comment with a link to https://github.com/whatwg/html/issues/490 and add a check for about:blank in the iframe load event handler and ignore the event if it matches.

CYBAI replied:
@jdm thanks a lot for helping me investigate this! That's clear for me now that why the test failed; I think we can only update the test and remove the comment once the issue is fixed.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #22648) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #22399) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #22732) made this pull request unmergeable. Please resolve the merge conflicts.

jdm reviewed (approved):
This looks good apart from the change to the test that I suggested!

CYBAI committed:
Make a workaround for the test with about:blank onload

As jdm mentioned in https://github.com/servo/servo/pull/22660#discussion_r249597546, the initial about:blank load event can be fired before the form navigation occurs.

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/web-platform-tests/wpt/pull/14995.

CYBAI commented:
@bors-servo try=wpt

jdm commented on tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/constructing-form-data-set.html:
```diff
@@ -103,6 +103,8 @@
   });
   let iframe = form.previousSibling;
   iframe.onload = t1.step_func_done(() => {
```
This will cause the test to finish after the about:blank load event. This should be `step_func` and there should be a `t1.done()` after the assert instead.

CYBAI replied:
Ahhh, yes! Thanks for the good catch! I only check it passed 😓 Let me update!

jdm commented on tests/wpt/web-platform-tests/html/semantics/forms/form-submission-0/constructing-form-data-set.html:
```diff
@@ -103,6 +103,8 @@
   });
   let iframe = form.previousSibling;
   iframe.onload = t1.step_func_done(() => {
+    // FIXME: https://github.com/whatwg/html/issues/490
```
I think a more clear comment would be:
```
// The initial about:blank load event can be fired before the form navigation occurs.
// See https://github.com/whatwg/html/issues/490 for more information.
```

CYBAI replied:
Thanks! This is indeed much more clear!

bors-servo commented:
:hourglass: Trying commit 9f54d76e6d416ca01c657f16c89855a3138f98c5 with merge 895ba0d78517665bbed0f0224f4895070799d4f6...

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#14995.

jdm commented:
@bors-servo delegate+
This needs `./mach update-manifest`, then it can merge.

bors-servo commented:
:v: @CYBAI can now approve this pull request

CYBAI commented:
@jdm Thanks! Waiting #22736 merged! I'll rebase and `update-manifest`!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #22736) made this pull request unmergeable. Please resolve the merge conflicts.

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at w3c/web-platform-tests#14995.

CYBAI commented:
@bors-servo r=jdm

CYBAI commented:
@bors-servo ping 🤔

bors-servo commented:
:sleepy: I'm awake I'm awake

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 3b9ab34 has been approved by `jdm`

<!-- @bors-servo r=jdm 3b9ab3460029b15fc0979883fa8523443851f713 -->

bors-servo commented:
:hourglass: Testing commit 3b9ab3460029b15fc0979883fa8523443851f713 with merge 461bce12415f8cddade2420c589fc2f669fc65d2...

jdm commented:
Looks like the webhook just missed the previous notification, since this PR did not show up in the homu queue.

CYBAI commented:
Maybe I should wait a minute to ask homu to approve after I force pushed 😂 Thanks for the approval!

bors-servo commented:
:sunny: Test successful - [android-mac](http://build.servo.org/builders/android-mac/builds/1002), [arm32](http://build.servo.org/builders/arm32/builds/13875), [arm64](http://build.servo.org/builders/arm64/builds/13903), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11508), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11564), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/6480), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/6447), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/10695), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/10755), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/6505), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/6435), [magicleap](http://build.servo.org/builders/magicleap/builds/621), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/BsX232ueR22Ep5NgXxDeAw)
Approved by: jdm
Pushing 461bce12415f8cddade2420c589fc2f669fc65d2 to master...
